subsys/bluetooth/host/classic/a2dp.c

```diff
@@ -76,35 +118,691 @@ struct bt_a2dp *get_new_connection(struct bt_conn *conn)
+		struct bt_a2dp_ep *local_ep, struct bt_a2dp_ep *remote_ep,
+		struct bt_a2dp_codec_cfg *config)
+{
+	if ((a2dp == NULL) || (stream == NULL) || (local_ep == NULL) ||
+		(remote_ep == NULL) || (config == NULL)) {
+		return -EINVAL;
+	}
+
+	stream->local_ep = local_ep;
+	stream->remote_ep = remote_ep;
+	stream->remote_ep_id = remote_ep->seid.sep.id;
+	stream->a2dp = a2dp;
+	local_ep->stream = stream;
+	remote_ep->stream = stream;
+	a2dp->set_config_param.req.func = bt_a2dp_set_config_cb;
+	a2dp->set_config_param.acp_stream_ep_id = remote_ep->seid.sep.id;
+	a2dp->set_config_param.int_stream_endpoint_id = local_ep->seid.sep.id;
+	a2dp->set_config_param.media_type = BT_AVDTP_AUDIO;
+	a2dp->set_config_param.media_codec_type = local_ep->codec_type;
+	a2dp->set_config_param.codec_specific_ie_len = config->codec_config->len;
+	a2dp->set_config_param.codec_specific_ie = &config->codec_config->codec_ie[0];
+	a2dp->set_config_param.seid = &local_ep->seid;
+	return bt_avdtp_set_configuration(&a2dp->session, &a2dp->set_config_param);
+}
+
+int bt_a2dp_stream_establish(struct bt_a2dp_stream *stream)
+{
+	struct bt_a2dp *a2dp;
+
+	if ((stream == NULL) || (stream->local_ep == NULL) || (stream->a2dp == NULL)) {
+		return -EINVAL;
+	}
+
+	a2dp = stream->a2dp;
+	a2dp->open_param.req.func = bt_a2dp_open_cb;
+	a2dp->open_param.acp_stream_ep_id = stream->remote_ep == NULL ?
+		stream->remote_ep->seid.sep.id : stream->remote_ep_id;
+	a2dp->open_param.seid = a2dp->set_config_param.seid;
+	return bt_avdtp_open(&a2dp->session, &a2dp->open_param);
+}
+
+int bt_a2dp_stream_release(struct bt_a2dp_stream *stream)
+{
+	/* todo: see the API description in a2dp.h */
+	return -EINVAL;
+}
+
+int bt_a2dp_stream_start(struct bt_a2dp_stream *stream)
+{
+	struct bt_a2dp *a2dp;
+
+	if ((stream == NULL) || (stream->local_ep == NULL) || (stream->a2dp == NULL)) {
+		return -EINVAL;
+	}
+
+	a2dp = stream->a2dp;
+	a2dp->start_param.req.func = bt_a2dp_start_cb;
+	a2dp->start_param.acp_stream_ep_id = stream->remote_ep == NULL ?
+		stream->remote_ep->seid.sep.id : stream->remote_ep_id;
+	a2dp->start_param.seid = &stream->local_ep->seid;
+	return bt_avdtp_start(&a2dp->session, &a2dp->start_param);
+}
+
+int bt_a2dp_stream_suspend(struct bt_a2dp_stream *stream)
+{
+	/* todo: see the API description in a2dp.h */
+	return -EINVAL;
+}
+
+int bt_a2dp_stream_reconfig(struct bt_a2dp_stream *stream, struct bt_a2dp_codec_cfg *config)
+{
+	/* todo: see the API description in a2dp.h */
+	return -EINVAL;
+}
+
+uint32_t bt_a2dp_get_mtu(struct bt_a2dp_stream *stream)
+{
+	if ((stream == NULL) || (stream->local_ep == NULL)) {
+		return -EPERM;
+	}
+
+	return bt_avdtp_get_media_mtu(&stream->local_ep->seid);
+}
+
+#if defined(CONFIG_BT_A2DP_SOURCE)
+int bt_a2dp_stream_send(struct bt_a2dp_stream *stream,  struct net_buf *buf,
+			uint16_t seq_num, uint32_t ts)
+{
+	struct bt_avdtp_media_hdr *media_hdr;
+
+	if (stream == NULL) {
+		return -EPERM;
```

Comment:
Change `-EPERM` to `-EINVAL`.

For all cases that the parameter is invalid, it is better to return `-EINVAL`.

Reply:
Updated